pallets/transfer-allowlist/src/lib.rs

```diff
@@ -0,0 +1,481 @@
+// Copyright 2023 Centrifuge Foundation (centrifuge.io).
+//
+// This file is part of the Centrifuge chain project.
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+//! This module checks whether an account should be allowed to make a transfer to
+//! a receiving location with a specific currency.
+//! If there are no allowances specified, then the account is assumed to be allowed
+//! to send to any location without restrictions.
+//! However once an allowance for a sender to a specific recieving location and currency is made,
+//! /then/ transfers from the sending account are restricted for that currency to:
+//! - the account(s) for which allowances have been made
+//! - the block range specified in the allowance
+use cfg_traits::TransferAllowance;
+use cfg_types::locations::Location;
+pub use pallet::*;
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use core::fmt::Debug;
+
+	use cfg_traits::ops::EnsureSub;
+	use codec::{Decode, Encode, MaxEncodedLen};
+	use frame_support::{
+		pallet_prelude::{DispatchResult, OptionQuery, StorageDoubleMap, StorageNMap, *},
+		traits::{tokens::AssetId, Currency, ReservableCurrency},
+		Twox64Concat,
+	};
+	use frame_system::pallet_prelude::{OriginFor, *};
+	use sp_runtime::{traits::AtLeast32BitUnsigned, Saturating};
+
+	use super::*;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Id type of Currency transfer restrictions will be checked for
+		type CurrencyId: AssetId
+			+ Parameter
+			+ Debug
+			+ Member
+			+ Copy
+			+ MaybeSerializeDeserialize
+			+ Ord
+			+ TypeInfo
+			+ MaxEncodedLen;
+
+		/// Deposit Balance to reserve/release
+		#[pallet::constant]
+		type Deposit: Get<DepositBalanceOf<Self>>;
+
+		/// Currency for Reserve/Unreserve with allowlist adding/removal,
+		/// given that the allowlist will be in storage
+		type ReserveCurrency: Currency<Self::AccountId> + ReservableCurrency<Self::AccountId>;
+	}
+
+	pub type DepositBalanceOf<T> = <<T as Config>::ReserveCurrency as Currency<
+		<T as frame_system::Config>::AccountId,
+	>>::Balance;
+
+	//
+	// Storage
+	//
+	pub type AllowanceDetailsOf<T> = AllowanceDetails<<T as frame_system::Config>::BlockNumber>;
```

Comment:
FYI We usually place these type aliases at the top, before `pub trait Config: frame_system::Config {`

Reply:
👍 fixed